build/v1/types.go

```diff
@@ -1335,3 +1348,64 @@ type SecretSpec struct {
 	// mountPath is the path at which to mount the secret
 	MountPath string `json:"mountPath" protobuf:"bytes,2,opt,name=mountPath"`
 }
+
+// BuildVolume describes a volume that is made available to build pods,
+// such that it can be mounted into buildah's runtime environment.
+// Only a subset of Kubernetes Volume sources are supported.
+type BuildVolume struct {
+	// name is a unique identifier for this BuildVolume.
+	// It must conform to the Kubernetes DNS label standard and be unique within the pod.
+	// Names that collide with those added by the build controller will result in a
+	// failed build with an error message detailing which name caused the error.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// source represents the location and type of the mounted volume.
+	Source BuildVolumeSource `json:"source" protobuf:"bytes,2,opt,name=source"`
+
+	// mounts represents the location of the volume in the image build container
+	Mounts []BuildVolumeMount `json:"mounts" protobuf:"bytes,3,opt,name=mounts"`
+}
+
+// BuildVolumeSourceType represents a build volume source type
+type BuildVolumeSourceType string
+
+const (
+	// BuildVolumeSourceTypeSecret is the Secret build source volume type
+	BuildVolumeSourceTypeSecret BuildVolumeSourceType = "Secret"
+
+	// BuildVolumeSourceTypeConfigmap is the ConfigMap build source volume type
+	BuildVolumeSourceTypeConfigMap BuildVolumeSourceType = "ConfigMap"
+)
+
+// BuildVolumeSource represents the source of a volume to mount
+// Only one of its supported types may be specified at any given time.
+type BuildVolumeSource struct {
+
+	// type is the BuildVolumeSourceType for the volume source.
+	// Type must match the populated volume source, and if not specified will be inferred on admission.
```

Comment:
what's the point in even having this if we're going to infer it?

Reply:
I'm not sure, @adambkaplan might have a better idea since he wrote the ep.

Reply:
This object is an example of a Union object, as describe by the upstream api conventions [1].
Scanning the various types of union objects upstream, it seems that some, but not all, objects employ the _type discriminator_ pattern, which is what I've advocated for here.
For instance, in the `Deployment` object, a union with a type discriminator is used for the deployment strategy [2].
However, type discriminator usage is not universal, most notably with the VolumeSource object.
It isn't clear if this is intentional or evidence of tech debt that can't be paid off.

My understanding, based on discussions with @deads2k, is that type discriminators with union objects help controllers in version skew situations. As an example, we can introduce a `PVC` volume type in v4.10. During an upgrade, we could have a cluster state where a 4.10 apiserver allows PVC volumes for builds, but the controller remains on 4.9 and can't start the build. The `Type` field lets us give a cleaner error message ("volume type PVC is not supported") versus a more vague and perhaps unfriendly error message (ex: "no volume source information was specified").

@deads2k @sttts I don't see any any recommendations from sig-architecture regarding type discriminator fields. Is this pattern something that is recommended upstream, something that we emphasize for OpenShift, or neither (i.e. this practice is outdated and we shouldn't do it)?

[1] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions
[2] https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/deployment-v1/#DeploymentSpec

Reply:
i'm less concerned with having it or not having and more concerned w/ it being optional.

either have it and make users specify it, or don't have it and handle the migration challenges in code, would be my expectation.

having it but having it be optional means it's not providing anything you can't already determine by looking at which other fields are populated.

Reply:
Making the Type required makes sense for this API.  We allow it to be optional in some cases in our configuration APIs and ascribe a "choose a good default for me which may change over time".  In this case, there is no reasonable default which can be chosen: additional information is always required for a functional API.

Both the error message and the clear intent are valuable for validation and client skew.  Using a Type field for unions is recommended.

Reply:
updated in next push.